### PR TITLE
transformers no longer ignore UnApply.fun

### PR DESCRIPTION
**Pull request by xeno-by**
Second time's the charm. I remember trying to do exactly the same somewhen
around 2.10.0-M4, but then some continuations tests were failing.
Luckily, today everything went smoothly.

Please note that this fix changes the way that SI-5465 manifests itself.
Previously it produced type errors, now it simply crashes the compiler.
Therefore I had to attach the try/catch FatalError clause to invocations
of toolbox methods, so that compiler crashes get caught and translated to
ToolBoxErrors.

Also fixes SI-7871, and that clears the way for implementing quasiquotes
with conventional macros rather than relying on a special case in typer.

### CODE REVIEW

**Comment by xeno-by** [comment]
review @retronym

**Comment by retronym** [comment]
Maybe its a bit outside the scope of this patch, but seeing as though we're in the area, I would prefer this sort of design:

```
object withCompiler {
  private lazy val global = { ... }
  private def cleanup() { ... }
  def apply[A](f: ToolBoxGlobal => A): A = try {
    f(global)
  } catch {
     case _: FatalError => ... 
  } finally {
     cleanup()
  }
}

...
withCompiler { compiler =>

}
```

I'm assuming it's cheap and correct to do the cleanup in all cases, even if not strictly necessary like in `parse`.

This way you can't forget to use the wrappers.

**Comment by xeno-by** [comment]
@retronym Done

File: src/compiler/scala/tools/reflect/ToolBoxFactory.scala [hunk 1]
```diff
@@ -28,6 +31,13 @@ abstract class ToolBoxFactory[U <: JavaUniverse](val u: U) { factorySelf =>
     lazy val classLoader = new AbstractFileClassLoader(virtualDirectory, factorySelf.mirror.classLoader)
     lazy val mirror: u.Mirror = u.runtimeMirror(classLoader)
 
+    lazy val arguments = CommandLineParser.tokenize(options)
+    lazy val virtualDirectory =
+      arguments.iterator.sliding(2).collectFirst{ case Seq("-d", dir) => dir }.lastOption match {
```
Comment (retronym) [inline src/compiler/scala/tools/reflect/ToolBoxFactory.scala]: `collectFirst` returns `Option`

File: src/compiler/scala/tools/reflect/ToolBoxFactory.scala [hunk 2]
```diff
@@ -28,6 +31,13 @@ abstract class ToolBoxFactory[U <: JavaUniverse](val u: U) { factorySelf =>
...
+        case Some(outDir) => scala.tools.nsc.io.AbstractFile.getDirectory(outDir)
```
Comment (retronym) [inline src/compiler/scala/tools/reflect/ToolBoxFactory.scala]: Add an import for this for balance with `VirtualDirectory`.

File: src/compiler/scala/tools/reflect/ToolBoxFactory.scala (hunk 1 above)
Comment (xeno-by) [inline src/compiler/scala/tools/reflect/ToolBoxFactory.scala]: Hahaha

**Comment by retronym** [comment]
LGTM

**Pull request merged**